Content.Server/StationEvents/Events/MeteorSwarm.cs

```diff
@@ -0,0 +1,126 @@
+
+        public override int EarliestStart => 30;
+        public override float Weight => WeightLow;
+        public override int? MaxOccurrences => 2;
+        public override int MinimumPlayers => 20;
+
+        public override string StartAnnouncement => "Meteors are on a collision course with the station. Brace for impact.";
+        protected override string EndAnnouncement => "The meteor swarm has passed. Please return to your stations.";
+
+        public override string? StartAudio => "/Audio/Announcements/bloblarm.ogg";
+
+        protected override float StartAfter => 30f;
+        protected override float EndAfter => float.MaxValue;
+
+        private float _cooldown;
+
+        /// <summary>
+        /// We'll send a specific amount of waves of meteors towards the station per ending rather than using a timer.
+        /// </summary>
+        private int _waveCounter;
+
+        private const int MinimumWaves = 3;
+        private const int MaximumWaves = 8;
+
+        private const float MinimumCooldown = 10f;
+        private const float MaximumCooldown = 60f;
+
+        private const int MeteorsPerWave = 5;
+        private const float MeteorVelocity = 10f;
+        private const float MaxAngularVelocity = 0.25f;
+        private const float MinAngularVelocity = -0.25f;
+
+        public override void Startup()
+        {
+            base.Startup();
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+            _waveCounter = robustRandom.Next(MinimumWaves, MaximumWaves);
+        }
+
+        public override void Shutdown()
+        {
+            base.Shutdown();
+            _waveCounter = 0;
+            _cooldown = 0f;
+            EndAfter = float.MaxValue;
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            if (!Started) return;
+
+            if (_waveCounter <= 0)
+            {
+                EndAfter = float.MinValue;
+            }
+            _cooldown -= frameTime;
+
+            if (_cooldown > 0f) return;
+
+            _waveCounter--;
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+
+            _cooldown += (MaximumCooldown - MinimumCooldown) * robustRandom.NextFloat() + MinimumCooldown;
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            var compManager = IoCManager.Resolve<IComponentManager>();
+
+            Box2? playableArea = null;
+            var mapId = EntitySystem.Get<GameTicker>().DefaultMap;
+
+            foreach (var grid in mapManager.GetAllGrids())
+            {
+                if (grid.ParentMapId != mapId || !compManager.TryGetComponent(grid.GridEntityId, out PhysicsComponent? gridBody)) continue;
+                var aabb = gridBody.GetWorldAABB();
+                playableArea ??= aabb;
+                playableArea = playableArea.Value.Union(aabb);
```

Comment:
what is going on here. What does union do? not saying its wrong but my codin' sense is tingling

Reply:
It gets the minimum box that covers both of them. It's also what physics AABB uses as well.

Reply:
i dont know how expensive union is but wouldn't it be better to
```suggestion
                playableArea = playableArea == null ? aabb : playableArea.Value.Union(aabb);
```

to cut the unneccesary union call when playableArea is null? I dont know how often it'd happen.

Reply:
It's just the Box2 equivalent of doing a Min / Max call, and given it's happening once every n seconds it won't even show on the profiler but sure.